File: solutions/python/arrays-hashing/design-hashset.py
# LeetCode 705 — Design HashSet (Easy)
# Category: Arrays & Hashing · Approach: Bucket Chaining
# Time: O(1) average | Space: O(n)
# Source: https://leetcode.com/problems/design-hashset/

class MyHashSet:
    def __init__(self) -> None:
        self.buckets: list[list[int]] = [[] for _ in range(1000)]

    def _bucket(self, key: int) -> list[int]:
        return self.buckets[key % 1000]

    def add(self, key: int) -> None:
        b = self._bucket(key)
        if key not in b:
            b.append(key)

    def remove(self, key: int) -> None:
        b = self._bucket(key)
        if key in b:
            b.remove(key)

    def contains(self, key: int) -> bool:
        return key in self._bucket(key)

File: solutions/typescript/arrays-hashing/design-hashset.ts
// LeetCode 705 — Design HashSet (Easy)
// Category: Arrays & Hashing · Approach: Bucket Chaining
// Time: O(1) average | Space: O(n)
// Source: https://leetcode.com/problems/design-hashset/

class MyHashSet {
  private buckets: number[][] = Array.from({ length: 1000 }, () => []);

  private bucket(key: number): number[] {
    return this.buckets[key % 1000];
  }

  add(key: number): void {
    const b = this.bucket(key);
    if (!b.includes(key)) b.push(key);
  }

  remove(key: number): void {
    const b = this.bucket(key);
    const i = b.indexOf(key);
    if (i !== -1) b.splice(i, 1);
  }

  contains(key: number): boolean {
    return this.bucket(key).includes(key);
  }
}
